Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,256 @@
+    }
+
+    dataLogLnIf(WaiterListsManagerInternal::verbose, "WaiterListManager notified waiters (count ", notified, ") for ptr ", RawPointer(ptr));
+    return notified;
+}
+
+void WaiterListManager::notifyWaiterImpl(const AbstractLocker&, Waiter* waiter, const ResolveResult resolveResult)
+{
+    ASSERT(waiter);
+
+    if (waiter->isAsync()) {
+        VM& vm = *waiter->getVM();
+        auto ticket = waiter->getTicket();
+        vm.deferredWorkTimer->scheduleWorkSoon(ticket, [resolveResult](DeferredWorkTimer::Ticket ticket) {
+            JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+            JSGlobalObject* globalObject = promise->globalObject();
+            VM& vm = promise->vm();
+            JSValue result = resolveResult == ResolveResult::Ok ? vm.smallStrings.okString() : vm.smallStrings.timedOutString();
+            promise->resolve(globalObject, result);
+        });
+
+        if (resolveResult == ResolveResult::Ok)
+            waiter->cancelTimer();
+        return;
+    }
+
+    waiter->nullOutVM();
+    waiter->getCondition().notifyOne();
+}
+
+size_t WaiterListManager::waiterListSize(void* ptr)
+{
+    WaiterList* list = findList(ptr);
+    size_t size = 0;
+    if (list) {
+        Locker locker { list->lock };
+        size = list->size();
+    }
+    return size;
+}
+
+void WaiterListManager::unregisterVM(VM* vm)
+{
+    Locker waiterListsLocker { m_waiterListsLock };
+    for (auto& entry : m_waiterLists) {
+        WaiterList* list = entry.value.get();
+        if (list) {
+            Locker locker { list->lock };
+            list->removeIf(locker, [&](Waiter* waiter) {
+                if (waiter->getVM() == vm) {
+                    dataLogLnIf(WaiterListsManagerInternal::verbose,
+                        "WaiterListManager::unregisterVM ",
+                        (waiter->isAsync() ? " deleted AsyncWaiter " : " removed SyncWaiter "),
+                        RawPointer(waiter),
+                        " in WaiterList for ptr ",
+                        RawPointer(entry.key));
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+}
+
+void WaiterListManager::unregisterSharedArrayBuffer(uint8_t* arrayPtr, size_t size)
+{
+    Locker locker { m_waiterListsLock };
+    m_waiterLists.removeIf([&](auto& entry) {
+        if (entry.key >= arrayPtr && entry.key < arrayPtr + size) {
+            WaiterList* list = entry.value.get();
+            if (list) {
+                Locker locker { list->lock };
+                list->removeIf(locker, [&](Waiter* waiter) {
+                    // If the AsyncWaiter has valid timer, then let it timeout. Otherwise un-task it.
+                    if (waiter->isAsync()) {
+                        if (waiter->hasTimer())
+                            return false;
+
+                        VM& vm = *waiter->getVM();
+                        {
+                            JSLockHolder locker(vm);
```

Comment:
This looks like it can cause deadlock.
We should not take a lock of VM. We should do similar thing to `scheduleWorkSoon`.